src/crypto/openssl.rs

```diff
@@ -0,0 +1,941 @@
+                let ecdsa_sig = EcdsaSig::from_private_components(r, s)
+                    .map_err(|_| AlgorithmError::InvalidData)?;
+                ecdsa_sig
+                    .verify(hash.as_ref(), public_key)
+                    .map_err(|_| AlgorithmError::InvalidData)?
+            }
+        };
+        if valid {
+            Ok(())
+        } else {
+            Err(AlgorithmError::BadSig)
+        }
+    }
+
+    /// Convert to a [`Dnskey`].
+    pub fn dnskey(&self) -> Dnskey<Vec<u8>> {
+        match self {
+            PublicKey::Default(message_digest, public_key, flags) => {
+                let alg = if *message_digest == MessageDigest::sha256() {
+                    SecAlg::RSASHA256
+                } else {
+                    unreachable!();
+                };
+                let rsa = public_key.rsa().expect("should not fail");
+                let e = rsa.e().to_vec();
+                let n = rsa.n().to_vec();
+
+                let mut key = Vec::new();
+
+                // Encode the exponent length.
+                if let Ok(exp_len) = u8::try_from(e.len()) {
+                    key.reserve_exact(1 + e.len() + n.len());
+                    key.push(exp_len);
+                } else if let Ok(exp_len) = u16::try_from(e.len()) {
+                    key.reserve_exact(3 + e.len() + n.len());
+                    key.push(0u8);
+                    key.extend(&exp_len.to_be_bytes());
+                } else {
+                    unreachable!(
+                        "RSA exponents are (much) shorter than 64KiB"
+                    )
+                }
+
+                key.extend(&*e);
+                key.extend(&*n);
+
+                Dnskey::new(*flags, 3, alg, key).expect("should not fail")
+            }
+            PublicKey::NoDigest(public_key, flags) => {
+                let alg = match public_key.id() {
+                    Id::ED25519 => SecAlg::ED25519,
+                    Id::ED448 => SecAlg::ED448,
+                    _ => todo!(),
+                };
+
+                let key =
+                    public_key.raw_public_key().expect("should not fail");
+                Dnskey::new(*flags, 3, alg, key).expect("should not fail")
+            }
+            PublicKey::EcDsa(message_digest, public_key, flags) => {
+                let alg = if *message_digest == MessageDigest::sha256() {
+                    SecAlg::ECDSAP256SHA256
+                } else if *message_digest == MessageDigest::sha384() {
+                    SecAlg::ECDSAP384SHA384
+                } else {
+                    unreachable!();
+                };
+
+                let key = public_key.public_key();
+                let group = public_key.group();
+                let mut ctx = BigNumContext::new().expect("should not fail");
+                let key = key
+                    .to_bytes(
+                        group,
+                        PointConversionForm::UNCOMPRESSED,
```

Comment:
Adding a reference here to RFC 6605 section 4 "DNSKEY and RRSIG Resource Records for ECDSA" which defines the byte format of teh key vec as being the "uncompressed form" would be nice here to explain the use of `PointConversionForm::UNCOMPRESSED` in the call to the OpenSSL `to_bytes()` function.